File: Sources/Theme/Theme.swift
import SwiftUI

// MARK: - Color Hex Extension

extension Color {
    init(hex: String) {
        let hex = hex.trimmingCharacters(in: CharacterSet.alphanumerics.inverted)
        var int: UInt64 = 0
        Scanner(string: hex).scanHexInt64(&int)
        let a, r, g, b: UInt64
        switch hex.count {
        case 6: // RGB
            (a, r, g, b) = (255, int >> 16, int >> 8 & 0xFF, int & 0xFF)
        case 8: // ARGB
            (a, r, g, b) = (int >> 24, int >> 16 & 0xFF, int >> 8 & 0xFF, int & 0xFF)
        default:
            (a, r, g, b) = (255, 0, 0, 0)
        }
        self.init(
            .sRGB,
            red: Double(r) / 255,
            green: Double(g) / 255,
            blue: Double(b) / 255,
            opacity: Double(a) / 255
        )
    }
}

// MARK: - NSColor Hex Extension

extension NSColor {
    convenience init(hex: String) {
        let hex = hex.trimmingCharacters(in: CharacterSet.alphanumerics.inverted)
        var int: UInt64 = 0
        Scanner(string: hex).scanHexInt64(&int)
        let r, g, b: UInt64
        switch hex.count {
        case 6:
            (r, g, b) = (int >> 16, int >> 8 & 0xFF, int & 0xFF)
        default:
            (r, g, b) = (0, 0, 0)
        }
        self.init(
            srgbRed: CGFloat(r) / 255,
            green: CGFloat(g) / 255,
            blue: CGFloat(b) / 255,
            alpha: 1.0
        )
    }
}

// MARK: - Theme

/// GitPilot 主题系统 — 严格按照 gitkraken-clone-ui-spec.md §1 §2 定义
enum Theme {
    // MARK: §1 配色 Token

    /// 应用最底层背景
    static let bgApp = Color(hex: "#16181d")
    /// 各面板背景
    static let bgPanel = Color(hex: "#1b1e24")
    /// 悬浮/选中行背景
    static let bgElevated = Color(hex: "#272a31")
    /// hover 态背景
    static let bgHover = Color(hex: "#22252c")
    /// 分隔线/边框
    static let border = Color(hex: "#2c2f36")
    /// 输入框等较明显边框
    static let borderStrong = Color(hex: "#3a3e46")
    /// 主文字
    static let textPrimary = Color(hex: "#dfe2e7")
    /// 次要/说明文字
    static let textSecondary = Color(hex: "#8b909a")
    /// 更弱的文字(占位符等)
    static let textMuted = Color(hex: "#646a73")
    /// 新增/已暂存/成功(+)
    static let accentGreen = Color(hex: "#4caf50")
    /// 主操作按钮底色(暗绿,禁用态)
    static let accentGreenDeep = Color(hex: "#2f5d3f")
    /// 主提交按钮可用态(亮绿)
    static let commitGreen = Color(hex: "#3c8c52")
    /// 分支标签 main 的色调
    static let accentTeal = Color(hex: "#3fb6a8")
    /// merge 连线 / 蓝色节点
    static let accentBlue = Color(hex: "#4a90e2")
    /// 删除按钮 / removed 行
    static let accentRed = Color(hex: "#c0392b")
    /// 当前激活标签背景
    static let selectedTab = Color(hex: "#23262d")
    /// 选中行背景
    static let selectedRow = Color(hex: "#23262d")
    /// 分支橙色
    static let accentOrange = Color(hex: "#e6994a")
    /// 进度提示黄色
    static let accentYellow = Color(hex: "#e9c54a")
    // MARK: NSColor versions (for NSTextView)

    static let nsBgApp = NSColor(hex: "#16181d")
    static let nsBgPanel = NSColor(hex: "#1b1e24")
    static let nsBgElevated = NSColor(hex: "#272a31")
    static let nsTextPrimary = NSColor(hex: "#dfe2e7")
    static let nsTextSecondary = NSColor(hex: "#8b909a")
    static let nsTextMuted = NSColor(hex: "#646a73")
    static let nsAccentGreen = NSColor(hex: "#4caf50")
    static let nsAccentRed = NSColor(hex: "#c0392b")
    static let nsLineNumber = NSColor(hex: "#5b616a")

    // MARK: Diff 专用色 (spec §1)

    /// 新增行背景(added) — 半透明绿
    static let diffAddedBg = Color(red: 46/255, green: 160/255, blue: 67/255, opacity: 0.15)
    /// 新增行左侧标记栏
    static let diffAddedGutter = Color(red: 46/255, green: 160/255, blue: 67/255, opacity: 0.28)
    /// 删除行背景(removed) — 半透明红
    static let diffRemovedBg = Color(red: 192/255, green: 57/255, blue: 43/255, opacity: 0.18)
    /// 删除行左侧标记栏
    static let diffRemovedGutter = Color(red: 192/255, green: 57/255, blue: 43/255, opacity: 0.30)
    /// 行号
    static let diffLineNumber = Color(hex: "#5b616a")

    // NSColor Diff versions
    static let nsDiffAddedBg = NSColor(srgbRed: 46/255, green: 160/255, blue: 67/255, alpha: 0.15)
    static let nsDiffAddedGutter = NSColor(srgbRed: 46/255, green: 160/255, blue: 67/255, alpha: 0.28)
    static let nsDiffRemovedBg = NSColor(srgbRed: 192/255, green: 57/255, blue: 43/255, alpha: 0.18)
    static let nsDiffRemovedGutter = NSColor(srgbRed: 192/255, green: 57/255, blue: 43/255, alpha: 0.30)

    // MARK: §2 字体排版

    /// UI 文字 — 系统无衬线字体
    static let uiFont = Font.system(size: 13)
    /// 标签栏标题
    static let tabFont = Font.system(size: 13)
    /// 列表行文字
    static let listFont = Font.system(size: 13)
    /// 区段标题 (字重 600)
    static let sectionFont = Font.system(size: 13, weight: .semibold)
    /// 说明性副文字
    static let captionFont = Font.system(size: 12)
    /// 代码/Diff 等宽字体
    static let codeFont = Font.custom("SF Mono", size: 12.5)
    /// 代码 fallback
    static let codeFontFallback = Font.system(size: 12.5, design: .monospaced)
    /// 小标签字体
    static let smallLabel = Font.system(size: 11)

    // NSFont versions
    static let nsCodeFont: NSFont = {
        if let font = NSFont(name: "SF Mono", size: 12.5) {
            return font
        }
        return NSFont.monospacedSystemFont(ofSize: 12.5, weight: .regular)
    }()

    // MARK: §3 尺寸常量

    /// 顶部窗口 + 标签栏高度
    static let titleBarHeight: CGFloat = 44
    /// 工具栏高度
    static let toolbarHeight: CGFloat = 64
    /// 左侧边栏展开宽度
    static let sidebarExpandedWidth: CGFloat = 260
    /// 左侧边栏折叠宽度
    static let sidebarCollapsedWidth: CGFloat = 48
    /// 右侧暂存面板宽度
    static let stagingPanelWidth: CGFloat = 470
    /// 提交行高度
    static let commitRowHeight: CGFloat = 36
    /// 列头高度
    static let columnHeaderHeight: CGFloat = 32
    /// 圆形节点直径
    static let commitNodeDiameter: CGFloat = 22

    // MARK: 分支颜色调色板

    /// 分支颜色轮换（不含紫色）
    static let branchColors: [Color] = [
        Color(hex: "#3fb6a8"), // teal
        Color(hex: "#4a90e2"), // blue
        Color(hex: "#e6994a"), // orange
        Color(hex: "#e25555"), // red
        Color(hex: "#4caf50"), // green
        Color(hex: "#e2cf4a"), // yellow
        Color(hex: "#4ac4e2"), // cyan
        Color(hex: "#e24a8f"), // pink
    ]

    /// 获取分支颜色(按 index 轮换)
    static func branchColor(at index: Int) -> Color {
        branchColors[index % branchColors.count]
    }
}

File: src/theme.rs
//! Colors, sizes and the GTK stylesheet — port of `Theme/Theme.swift`.
//! Every color token from the macOS spec lives here; views only reference classes.
// The palette is intentionally complete even if a few tokens are unused so far.
#![allow(dead_code)]

// MARK: - Color tokens (hex)

pub const BG_APP: &str = "#16181d";
pub const BG_PANEL: &str = "#1b1e24";
pub const BG_ELEVATED: &str = "#272a31";
pub const BG_HOVER: &str = "#22252c";
pub const BORDER: &str = "#2c2f36";
pub const BORDER_STRONG: &str = "#3a3e46";
pub const TEXT_PRIMARY: &str = "#dfe2e7";
pub const TEXT_SECONDARY: &str = "#8b909a";
pub const TEXT_MUTED: &str = "#646a73";
pub const ACCENT_GREEN: &str = "#4caf50";
pub const ACCENT_GREEN_DEEP: &str = "#2f5d3f";
pub const COMMIT_GREEN: &str = "#3c8c52";
pub const ACCENT_TEAL: &str = "#3fb6a8";
pub const ACCENT_BLUE: &str = "#4a90e2";
pub const ACCENT_RED: &str = "#c0392b";
pub const SELECTED_ROW: &str = "#23262d";
pub const ACCENT_ORANGE: &str = "#e6994a";
pub const ACCENT_YELLOW: &str = "#e9c54a";
pub const DIFF_LINE_NUMBER: &str = "#5b616a";

/// Branch lane palette (no purple), matching the Mac app.
pub const BRANCH_COLORS: [&str; 8] = [
    "#3fb6a8", // teal
    "#4a90e2", // blue
    "#e6994a", // orange
    "#e25555", // red
    "#4caf50", // green
    "#e2cf4a", // yellow
    "#4ac4e2", // cyan
    "#e24a8f", // pink
];

pub fn branch_color_hex(index: usize) -> &'static str {
    BRANCH_COLORS[index % BRANCH_COLORS.len()]
}

pub fn branch_color_rgb(index: usize) -> (f64, f64, f64) {
    hex_to_rgb(branch_color_hex(index))
}

/// "#rrggbb" → (r, g, b) as 0.0..=1.0 floats for Cairo.
pub fn hex_to_rgb(hex: &str) -> (f64, f64, f64) {
    let h = hex.trim_start_matches('#');
    let r = u8::from_str_radix(h.get(0..2).unwrap_or("00"), 16).unwrap_or(0);
    let g = u8::from_str_radix(h.get(2..4).unwrap_or("00"), 16).unwrap_or(0);
    let b = u8::from_str_radix(h.get(4..6).unwrap_or("00"), 16).unwrap_or(0);
    (r as f64 / 255.0, g as f64 / 255.0, b as f64 / 255.0)
}

// MARK: - Sizes

pub const TITLE_BAR_HEIGHT: i32 = 44;
pub const TOOLBAR_HEIGHT: i32 = 58;
pub const SIDEBAR_EXPANDED_WIDTH: i32 = 260;
pub const SIDEBAR_COLLAPSED_WIDTH: i32 = 48;
/// Draggable floor for the sidebar. The pane opens at SIDEBAR_EXPANDED_WIDTH but
/// the user can pull the divider in to here (well below the old 260 hard stop).
pub const SIDEBAR_MIN_WIDTH: i32 = 150;
pub const STAGING_PANEL_WIDTH: i32 = 460;
/// Draggable floor for the right (staging / commit-detail) panel.
pub const STAGING_PANEL_MIN_WIDTH: i32 = 260;
pub const COLUMN_HEADER_HEIGHT: i32 = 32;

// Commit graph geometry
pub const GRAPH_ROW_HEIGHT: f64 = 40.0;
pub const GRAPH_LANE_SPACING: f64 = 18.0;
pub const GRAPH_LEADING_PAD: f64 = 16.0;
pub const GRAPH_NODE_RADIUS: f64 = 11.0;

// MARK: - Stylesheet

/// The full application stylesheet. Loaded once at startup.
pub const CSS: &str = r#"
/* ---- color variables ---- */
@define-color bg_app #16181d;
@define-color bg_panel #1b1e24;
@define-color bg_elevated #272a31;
@define-color bg_hover #22252c;
@define-color border_c #2c2f36;
@define-color border_strong #3a3e46;
@define-color text_primary #dfe2e7;
@define-color text_secondary #8b909a;
@define-color text_muted #646a73;
@define-color accent_green #4caf50;
@define-color commit_green #3c8c52;
@define-color accent_teal #3fb6a8;
@define-color accent_blue #4a90e2;
@define-color accent_red #c0392b;
@define-color accent_orange #e6994a;
@define-color accent_yellow #e9c54a;
@define-color selected_row #23262d;

* { outline: none; }

window, .app-bg { background-color: @bg_app; color: @text_primary; }

/* generic text tones */
.primary-text { color: @text_primary; }
.dim         { color: @text_secondary; }
.muted       { color: @text_muted; }
.teal        { color: @accent_teal; }
.orange      { color: @accent_orange; }
.blue        { color: @accent_blue; }
.green       { color: @accent_green; }
.red         { color: @accent_red; }

.panel     { background-color: @bg_panel; }
.elevated  { background-color: @bg_elevated; }

/* full-bleed bar: background reaches the edges, content inset via padding
   (a margin would leave gaps where the darker app background shows through) */
.bar-pad { padding-left: 14px; padding-right: 14px; }

.section-title { font-weight: 600; color: @text_secondary; font-size: 13px; }
.selector-title { font-weight: 600; color: @text_primary; font-size: 14px; }
.caption { font-size: 11px; color: @text_muted; }
.tiny { font-size: 10px; }
.h1 { font-size: 24px; font-weight: 600; color: @text_primary; }
.h2 { font-size: 17px; font-weight: 600; color: @text_primary; }

/* ---- borders as thin separators ---- */
separator { background-color: @border_c; min-width: 1px; min-height: 1px; }
.border-bottom { border-bottom: 1px solid @border_c; }
.border-top { border-top: 1px solid @border_c; }
.border-right { border-right: 1px solid @border_c; }
.border-left { border-left: 1px solid @border_c; }

/* ---- buttons ---- */
button {
    background-image: none;
    background-color: transparent;
    border: none;
    box-shadow: none;
    color: @text_secondary;
    min-height: 0;
    padding: 4px 6px;
    border-radius: 6px;
    text-shadow: none;
}
button:hover { background-color: @bg_hover; }
button:active { background-color: @bg_elevated; }
button:disabled { color: @text_muted; background-color: transparent; }

/* Toolbar action buttons: enabled ones read at full contrast (like the macOS
   toolbar) so they look clickable; only :disabled dims to muted. */
.toolbar-btn { border-radius: 7px; color: @text_primary; padding: 5px 8px; }
.toolbar-btn:hover { background-color: @bg_hover; }
.toolbar-btn:active { background-color: @bg_elevated; }
.toolbar-btn:disabled { color: @text_muted; background-color: transparent; }
.toolbar-btn label { color: inherit; }
.tb-label { font-size: 11px; font-weight: 500; }

.icon-btn { padding: 4px; border-radius: 4px; min-width: 24px; min-height: 24px; }

/* primary (commit) button */
.primary-btn {
    background-color: @commit_green;
    color: white;
    border-radius: 5px;
    padding: 9px 12px;
    font-weight: 600;
}
.primary-btn:hover { background-color: shade(@commit_green, 1.08); }
.primary-btn:active { background-color: shade(@commit_green, 0.92); }
.primary-btn:disabled { background-color: #2f5d3f; color: @text_muted; }

/* outline (Stage/Unstage) */
.outline-btn {
    color: @accent_green;
    background-color: alpha(@accent_green, 0.08);
    border: 1px solid alpha(@accent_green, 0.5);
    border-radius: 4px;
    padding: 3px 8px;
    font-size: 11px;
    font-weight: 500;
}
.outline-btn:hover { background-color: alpha(@accent_green, 0.16); border-color: alpha(@accent_green, 0.85); }
.outline-btn:active { background-color: alpha(@accent_green, 0.24); }

.danger-btn {
    background-color: @accent_red;
    color: white;
    border-radius: 6px;
    padding: 6px;
}
.danger-btn:hover { background-color: shade(@accent_red, 1.1); }

/* segmented control */
.segment { background-color: @bg_app; border-radius: 6px; padding: 2px; }
.segment button { border-radius: 5px; padding: 4px 12px; color: @text_secondary; font-size: 12px; font-weight: 500; }
.segment button:hover { background-color: @bg_hover; }
.segment button.active { background-color: @bg_elevated; color: @text_primary; }

/* ---- tabs (title bar) ---- */
.tab {
    padding: 0 4px;
    border-radius: 6px 6px 0 0;
    background-color: transparent;
}
.tab:hover { background-color: @bg_hover; }
.tab.active { background-color: @selected_row; }
.tab button { background: transparent; box-shadow: none; min-height: 0; padding: 2px 4px; }
.tab-select { padding: 2px 8px; }
.tab-select:hover { background: transparent; }
.tab-close { border-radius: 5px; padding: 3px; min-width: 18px; min-height: 18px; }
.tab-close:hover { background-color: alpha(@text_primary, 0.14); }

/* ---- list rows ---- */
.list-row { border-radius: 0; padding: 0; }
.row-hover:hover { background-color: @bg_hover; }
.row-selected { background-color: @selected_row; }

/* ---- entries ---- */
entry, .sgc-entry {
    background-color: @bg_app;
    color: @text_primary;
    border: 1px solid @border_strong;
    border-radius: 5px;
    padding: 7px 8px;
    box-shadow: none;
    background-image: none;
    min-height: 0;
    caret-color: @text_primary;
}
entry:focus, .sgc-entry:focus-within { border-color: @accent_teal; }
entry image { color: @text_muted; }
textview, textview text {
    background-color: @bg_app;
    color: @text_primary;
    caret-color: @text_primary;
}
.code text, .code {
    font-family: "Source Code Pro", "DejaVu Sans Mono", monospace;
    font-size: 12.5px;
}

/* checkbox */
checkbutton { color: @text_secondary; font-size: 12px; }
checkbutton check { background-color: @bg_app; border: 1px solid @border_strong; }
checkbutton check:checked { background-color: @commit_green; border-color: @commit_green; color: white; }

/* ---- pills / badges ---- */
.pill {
    border-radius: 999px;
    padding: 2px 7px;
    font-size: 11px;
    font-weight: 500;
    border: 1px solid transparent;
}
.pill-local  { color: @accent_teal;   background-color: alpha(@accent_teal, 0.12);   border-color: alpha(@accent_teal, 0.6); }
.pill-remote { color: @accent_orange; background-color: alpha(@accent_orange, 0.12); border-color: alpha(@accent_orange, 0.6); }
.pill-tag    { color: @accent_blue;   background-color: alpha(@accent_blue, 0.12);   border-color: alpha(@accent_blue, 0.6); }
.badge-green { color: @accent_green;  background-color: alpha(@accent_green, 0.15); border-radius: 999px; padding: 1px 6px; font-size: 11px; font-weight: 700; }
.badge-count { color: @text_primary;  background-color: @accent_teal; border-radius: 999px; padding: 0 6px; font-size: 9px; font-weight: 700; }
.branch-chip { color: @accent_teal; background-color: alpha(@accent_teal, 0.15); border-radius: 999px; padding: 2px 7px; font-size: 11px; font-weight: 600; }

/* ---- diff ---- */
.hunk-header { background-color: @bg_elevated; color: @text_muted; padding: 3px 12px; }
.diff-added   { background-color: alpha(#2ea043, 0.15); }
.diff-removed { background-color: alpha(#c0392b, 0.18); }
.gutter { color: #5b616a; }
.gutter-added   { background-color: alpha(#2ea043, 0.28); color: #5b616a; }
.gutter-removed { background-color: alpha(#c0392b, 0.30); color: #5b616a; }
.gutter-plain   { background-color: @bg_panel; color: #5b616a; }
.marker-added { color: @accent_green; }
.marker-removed { color: @accent_red; }

/* ---- toast / activity ---- */
.toast {
    border-radius: 9px;
    padding: 12px 16px;
    color: white;
    font-weight: 600;
    font-size: 13px;
}
.toast-success  { background-color: @accent_green; }
.toast-error    { background-color: @accent_red; }
.toast-info     { background-color: @accent_blue; }
.toast-activity { background-color: @accent_yellow; color: @bg_app; }

/* selected big header title in dialogs, etc. */
.card { background-color: @bg_panel; border: 1px solid @border_c; border-radius: 8px; }

scrollbar { background-color: transparent; }
scrollbar slider { background-color: @border_strong; border-radius: 999px; min-width: 6px; min-height: 6px; }
scrollbar slider:hover { background-color: @text_muted; }

/* progressbar / spinner tint */
spinner { color: @text_secondary; }

tooltip { background-color: @bg_elevated; color: @text_primary; border: 1px solid @border_c; }

popover contents, popover > arrow { background-color: @bg_elevated; color: @text_primary; border: 1px solid @border_c; }
popover button { color: @text_primary; }
popover button:hover { background-color: @bg_hover; }
"#;
